<a href="https://colab.research.google.com/github/FaN0016/UPISummerCourseAssisgment/blob/main/Course_2B.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Group 21**


https://colab.research.google.com/drive/1ihZrpvlJhfvEAM9mXhTKMLDeHlhk7hyP?usp=sharing

Exploring the data Sales involves a step-by-step process:

1. Check and prepare data to clean and handling missing values and ensuring consistency.
2. Summaries the data with statistical analysis: Use descriptive statistics with aggregation function (i.e sum, count, average, min, max) for searching meaningful information such as: top product sales, total amount, average amount, etc
3. Use Statistical methods to identify significant correlation/comparative/distribution/trending between variables from the data
4. Visualize the data with charts and graphs to see patterns and relationships (min.3 graph)
5. Use related python library to handle all of tasks
6. Upload your source code with python extension file such as .py or .ipynb and file .rawgraphs (if you visualized the data using rawgraphs)
7. Tomorrow some of you will present the result of your assignment

# First Steps (importing all requirement and make csv is real csv)

Import all library you need

In [12]:
import pandas as pd #
import numpy as np #
import matplotlib.pyplot as plt #
import seaborn as sns #

Import dataset
This code run with google colab so using this path as default. Change if needed

In [13]:
dataset_path = "/content/orderdataset.csv"
order_test_df = pd.read_csv(dataset_path)

test. If path working you can see the table

In [14]:
order_test_df.head()

order_id;quantity;product_id;price;seller_id;freight_value;customer_id;order_status;purchase_date;payment_type;product_category_name;product_weight_gram
0  2e7a8482f6fb09756ca50c10d7bfc047;2;f293394c72c...                                                                                                      
1  2e7a8482f6fb09756ca50c10d7bfc047;1;c1488892604...                                                                                                      
2  e5fa5a7210941f7d56d0208e4e071d35;1;f3c2d01a84c...                                                                                                      
3  3b697a20d9e427646d92567910af6d57;1;3ae08df6bcb...                                                                                                      
4  71303d7e93b399f5bcd537d124c0bcfa;1;d2998d7ced1...

as we can see here data is not normal format for csv (because is not separated by comma but with semicolon so we need to change that). We have 2 option (as far as we know)

1.   Just use sep=";" argument on pd.read_csv() so pandas will read semicolon just like read using comma (what we doing here is this) or,
2.   Convert first using to_csv() and just use new csv with comma as separator for code.



In [15]:
order_df = pd.read_csv(dataset_path, sep=";")

In [16]:
order_df.head()

order_id  quantity  \
0  2e7a8482f6fb09756ca50c10d7bfc047         2   
1  2e7a8482f6fb09756ca50c10d7bfc047         1   
2  e5fa5a7210941f7d56d0208e4e071d35         1   
3  3b697a20d9e427646d92567910af6d57         1   
4  71303d7e93b399f5bcd537d124c0bcfa         1   

                         product_id    price  \
0  f293394c72c9b5fafd7023301fc21fc2  1489000   
1  c1488892604e4ba5cff5b4eb4d595400  1756000   
2  f3c2d01a84c947b078e32bbef0718962  1707000   
3  3ae08df6bcbfe23586dd431c40bddbb7  3071000   
4  d2998d7ced12f83f9b832f33cf6507b6  3833000   

                          seller_id  freight_value  \
0  1554a68530182680ad5c8b042c3ab563          28000   
1  1554a68530182680ad5c8b042c3ab563          45000   
2  a425f92c199eb576938df686728acd20         174000   
3  522620dcb18a6b31cd7bdf73665113a9         154000   
4  25e6ffe976bd75618accfe16cefcbd0d         147000   

                        customer_id order_status purchase_date  \
0  08c5351a6aca1c1589a38f244edeee9d      shipped    05/09/2016   
1  08c5351a6aca1c1589a38f244edeee9d      shipped    05/09/2016   
2  683c54fc24d40ee9f8a6fc179fd9856c     canceled    05/09/2016   
3  355077684019f7f60a031656bd7262b8    delivered    03/10/2016   
4  b106b360fe2ef8849fbbd056f777b4d5     canceled    03/10/2016   

      payment_type product_category_name  product_weight_gram  
0  virtual account               fashion               1800.0  
1  virtual account            automotive               1400.0  
2         e-wallet                  toys                700.0  
3         e-wallet             utilities                300.0  
4         e-wallet               fashion                500.0

As now we can see here, we can read the table like normal table. Now we move to next step

# Data Cleaning

## Check data for cleaning

In [17]:
# check how much data in here
order_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49999 entries, 0 to 49998
Data columns (total 12 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   order_id               49999 non-null  object 
 1   quantity               49999 non-null  int64  
 2   product_id             49999 non-null  object 
 3   price                  49999 non-null  int64  
 4   seller_id              49999 non-null  object 
 5   freight_value          49999 non-null  int64  
 6   customer_id            49999 non-null  object 
 7   order_status           49999 non-null  object 
 8   purchase_date          49999 non-null  object 
 9   payment_type           49999 non-null  object 
 10  product_category_name  49999 non-null  object 
 11  product_weight_gram    49980 non-null  float64
dtypes: float64(1), int64(3), object(8)
memory usage: 4.6+ MB


As we can see here we have missing data. On product_weight_gram is just 49980 so missing like 19 data. Now we check for data duplication, we can clean that later.

In [18]:
print("Sum duplicate data : ", order_df.duplicated().sum())

Sum duplicate data :  0


We don't have any data is duplicated. So next is checking anomaly on data (like wierd input)

In [19]:
order_df.describe()

quantity         price  freight_value  product_weight_gram
count  49999.000000  4.999900e+04   49999.000000         49980.000000
mean       1.197484  2.607784e+06  104521.390428          2201.830892
std        0.722262  1.388312e+06   55179.844962          3929.896875
min        1.000000  2.000000e+05    9000.000000            50.000000
25%        1.000000  1.410500e+06   57000.000000           300.000000
50%        1.000000  2.610000e+06  104000.000000           800.000000
75%        1.000000  3.810000e+06  152000.000000          1850.000000
max       21.000000  5.000000e+06  200000.000000         40425.000000

Nothing wierd in here

## Cleaning data missing

Because we have 19 data missing we can check now what missing

In [20]:
order_df[order_df.product_weight_gram.isna()]

order_id  quantity  \
1193   85f8ad45e067abd694b627859fa57453         1   
5624   eb855beb3ac99461f7a076b4c3652472         1   
6284   a2456e7f02197951664897a94c87242d         1   
6312   e3daea0200104991cb979c2fcc509ae7         1   
6433   6f497c40431d5fb0cfbd6c943dd29215         1   
6480   1521c6bb7b1028154c8c67cf80fa809f         1   
6499   415cfaaaa8cea49f934470548797fed1         2   
6500   415cfaaaa8cea49f934470548797fed1         1   
6555   101157d4fae1c9fb74a00a5dee265c25         1   
7046   c1424efcde3c9e9febd9e1761667789e         1   
8452   ddf16d77e858a32f36e10c289a28ef61         1   
8453   ddf16d77e858a32f36e10c289a28ef61         1   
14008  d38dcb503cd4ddc6ce7702552918bd8f         2   
14009  d38dcb503cd4ddc6ce7702552918bd8f         1   
16301  6e150190fbe04c642a9cf0b80d83ee16         1   
17353  bf49f84a0580ef6751e13357776b7ed9         1   
18183  bbfc7badbed2f1828e22b6d629201bd4         1   
23218  595316a07cd3dea9db7adfcc7e247ae7         1   
25072  a7a43f469c0d7bdb0a23a82db125aefa         1   

                             product_id    price  \
1193   09ff539a621711667c43eba6a3bd8466  2899000   
5624   5eb564652db742ff8f28759cd8d2652a   772000   
6284   5eb564652db742ff8f28759cd8d2652a  2079000   
6312   5eb564652db742ff8f28759cd8d2652a  1379000   
6433   5eb564652db742ff8f28759cd8d2652a  3591000   
6480   5eb564652db742ff8f28759cd8d2652a  1425000   
6499   5eb564652db742ff8f28759cd8d2652a  4788000   
6500   5eb564652db742ff8f28759cd8d2652a  1069000   
6555   5eb564652db742ff8f28759cd8d2652a  2841000   
7046   5eb564652db742ff8f28759cd8d2652a   368000   
8452   5eb564652db742ff8f28759cd8d2652a  4580000   
8453   5eb564652db742ff8f28759cd8d2652a  4126000   
14008  5eb564652db742ff8f28759cd8d2652a  2540000   
14009  5eb564652db742ff8f28759cd8d2652a   264000   
16301  5eb564652db742ff8f28759cd8d2652a   565000   
17353  5eb564652db742ff8f28759cd8d2652a  3137000   
18183  5eb564652db742ff8f28759cd8d2652a  1382000   
23218  5eb564652db742ff8f28759cd8d2652a  2872000   
25072  5eb564652db742ff8f28759cd8d2652a  4380000   

                              seller_id  freight_value  \
1193   8b8cfc8305aa441e4239358c9f6f2485          45000   
5624   4e922959ae960d389249c378d1c939f5          48000   
6284   4e922959ae960d389249c378d1c939f5         160000   
6312   4e922959ae960d389249c378d1c939f5         179000   
6433   4e922959ae960d389249c378d1c939f5          87000   
6480   4e922959ae960d389249c378d1c939f5          89000   
6499   4e922959ae960d389249c378d1c939f5          67000   
6500   4e922959ae960d389249c378d1c939f5          36000   
6555   4e922959ae960d389249c378d1c939f5         111000   
7046   4e922959ae960d389249c378d1c939f5         150000   
8452   4e922959ae960d389249c378d1c939f5         104000   
8453   4e922959ae960d389249c378d1c939f5         145000   
14008  4e922959ae960d389249c378d1c939f5         112000   
14009  4e922959ae960d389249c378d1c939f5         100000   
16301  4e922959ae960d389249c378d1c939f5         189000   
17353  4e922959ae960d389249c378d1c939f5          21000   
18183  4e922959ae960d389249c378d1c939f5         154000   
23218  4e922959ae960d389249c378d1c939f5         123000   
25072  4e922959ae960d389249c378d1c939f5         200000   

                            customer_id order_status purchase_date  \
1193   1d088dea8732788ec35dd4ee6dd76112    delivered    30/01/2017   
5624   c91289ce43149a8ea5560d446f1d1dd2    delivered    23/03/2017   
6284   7317f41f2cf650174af819cdb68284f0    delivered    30/03/2017   
6312   4730251e8934a542a009d77dfd027375    delivered    30/03/2017   
6433   5beb36d1757aa17a044222a7d79b9820    delivered    01/04/2017   
6480   ca29b2bf57243228e98eab2dab805ae9    delivered    02/04/2017   
6499   a8dff6357fea30071032ff2091d16430    delivered    02/04/2017   
6500   a8dff6357fea30071032ff2091d16430    delivered    02/04/2017   
6555   f72b2f8d9295ef93fd40a4c49f67a42b    delivered    03/04/2017   
7046   8a80133b8ace6b21415367a131a75a26    delivered    07/04/2017  

We have some option to clean it

1.   We can just drop all NaN data and do all that without NaN data
2.   We can use some smart guess (using value we have) for filling NaN data we have.





### Trying smart guess

We can try using some smart guess. We have product_id in here. If product_id who have NaN value is same as product_id with value on product_weight_gram we can just cross reference and fill NaN value. We need to check first if we have another duplicate product_id

In [21]:
product_ids_with_nan = order_df[order_df.product_weight_gram.isna()]['product_id'].unique()
print(product_ids_with_nan)

['09ff539a621711667c43eba6a3bd8466' '5eb564652db742ff8f28759cd8d2652a']


In [22]:
order_df[order_df['product_id'] == '09ff539a621711667c43eba6a3bd8466']

order_id  quantity  \
1193  85f8ad45e067abd694b627859fa57453         1   

                            product_id    price  \
1193  09ff539a621711667c43eba6a3bd8466  2899000   

                             seller_id  freight_value  \
1193  8b8cfc8305aa441e4239358c9f6f2485          45000   

                           customer_id order_status purchase_date  \
1193  1d088dea8732788ec35dd4ee6dd76112    delivered    30/01/2017   

     payment_type product_category_name  product_weight_gram  
1193  credit card                sports                  NaN

In [23]:
order_df[order_df['product_id'] == '5eb564652db742ff8f28759cd8d2652a']

order_id  quantity  \
5624   eb855beb3ac99461f7a076b4c3652472         1   
6284   a2456e7f02197951664897a94c87242d         1   
6312   e3daea0200104991cb979c2fcc509ae7         1   
6433   6f497c40431d5fb0cfbd6c943dd29215         1   
6480   1521c6bb7b1028154c8c67cf80fa809f         1   
6499   415cfaaaa8cea49f934470548797fed1         2   
6500   415cfaaaa8cea49f934470548797fed1         1   
6555   101157d4fae1c9fb74a00a5dee265c25         1   
7046   c1424efcde3c9e9febd9e1761667789e         1   
8452   ddf16d77e858a32f36e10c289a28ef61         1   
8453   ddf16d77e858a32f36e10c289a28ef61         1   
14008  d38dcb503cd4ddc6ce7702552918bd8f         2   
14009  d38dcb503cd4ddc6ce7702552918bd8f         1   
16301  6e150190fbe04c642a9cf0b80d83ee16         1   
17353  bf49f84a0580ef6751e13357776b7ed9         1   
18183  bbfc7badbed2f1828e22b6d629201bd4         1   
23218  595316a07cd3dea9db7adfcc7e247ae7         1   
25072  a7a43f469c0d7bdb0a23a82db125aefa         1   

                             product_id    price  \
5624   5eb564652db742ff8f28759cd8d2652a   772000   
6284   5eb564652db742ff8f28759cd8d2652a  2079000   
6312   5eb564652db742ff8f28759cd8d2652a  1379000   
6433   5eb564652db742ff8f28759cd8d2652a  3591000   
6480   5eb564652db742ff8f28759cd8d2652a  1425000   
6499   5eb564652db742ff8f28759cd8d2652a  4788000   
6500   5eb564652db742ff8f28759cd8d2652a  1069000   
6555   5eb564652db742ff8f28759cd8d2652a  2841000   
7046   5eb564652db742ff8f28759cd8d2652a   368000   
8452   5eb564652db742ff8f28759cd8d2652a  4580000   
8453   5eb564652db742ff8f28759cd8d2652a  4126000   
14008  5eb564652db742ff8f28759cd8d2652a  2540000   
14009  5eb564652db742ff8f28759cd8d2652a   264000   
16301  5eb564652db742ff8f28759cd8d2652a   565000   
17353  5eb564652db742ff8f28759cd8d2652a  3137000   
18183  5eb564652db742ff8f28759cd8d2652a  1382000   
23218  5eb564652db742ff8f28759cd8d2652a  2872000   
25072  5eb564652db742ff8f28759cd8d2652a  4380000   

                              seller_id  freight_value  \
5624   4e922959ae960d389249c378d1c939f5          48000   
6284   4e922959ae960d389249c378d1c939f5         160000   
6312   4e922959ae960d389249c378d1c939f5         179000   
6433   4e922959ae960d389249c378d1c939f5          87000   
6480   4e922959ae960d389249c378d1c939f5          89000   
6499   4e922959ae960d389249c378d1c939f5          67000   
6500   4e922959ae960d389249c378d1c939f5          36000   
6555   4e922959ae960d389249c378d1c939f5         111000   
7046   4e922959ae960d389249c378d1c939f5         150000   
8452   4e922959ae960d389249c378d1c939f5         104000   
8453   4e922959ae960d389249c378d1c939f5         145000   
14008  4e922959ae960d389249c378d1c939f5         112000   
14009  4e922959ae960d389249c378d1c939f5         100000   
16301  4e922959ae960d389249c378d1c939f5         189000   
17353  4e922959ae960d389249c378d1c939f5          21000   
18183  4e922959ae960d389249c378d1c939f5         154000   
23218  4e922959ae960d389249c378d1c939f5         123000   
25072  4e922959ae960d389249c378d1c939f5         200000   

                            customer_id order_status purchase_date  \
5624   c91289ce43149a8ea5560d446f1d1dd2    delivered    23/03/2017   
6284   7317f41f2cf650174af819cdb68284f0    delivered    30/03/2017   
6312   4730251e8934a542a009d77dfd027375    delivered    30/03/2017   
6433   5beb36d1757aa17a044222a7d79b9820    delivered    01/04/2017   
6480   ca29b2bf57243228e98eab2dab805ae9    delivered    02/04/2017   
6499   a8dff6357fea30071032ff2091d16430    delivered    02/04/2017   
6500   a8dff6357fea30071032ff2091d16430    delivered    02/04/2017   
6555   f72b2f8d9295ef93fd40a4c49f67a42b    delivered    03/04/2017   
7046   8a80133b8ace6b21415367a131a75a26    delivered    07/04/2017   
8452   84cc013dd1790fdafb0fa598695cf3c3    delivered    24/04/2017   
8453   84cc013dd1790fdafb0fa598695cf3c3    delivered    24/04/2017   
14008  b0a3a02fe893d9a9385a98db1348244b    delivered    05/06/2017   
14009  b0a3a02fe893d9

as we can check we have 2 product_id who have NaN value and that 2 product_id is unique only for NaN value so we can just simply drop NaN value because we cannot check what item in there

### Just drop the table

In [24]:
order_df.dropna(subset=['product_weight_gram'], inplace=True)

Check data if still have data missing

In [25]:
order_df.isna().sum()

order_id                 0
quantity                 0
product_id               0
price                    0
seller_id                0
freight_value            0
customer_id              0
order_status             0
purchase_date            0
payment_type             0
product_category_name    0
product_weight_gram      0
dtype: int64

We don't have any data missing anymore

# Explanitory Data Analysis (EDA)

In [26]:
order_df.describe(include="all")

order_id      quantity  \
count                              49980  49980.000000   
unique                             42678           NaN   
top     8272b63d03f5f79c56e9e4120aec44ef           NaN   
freq                                  21           NaN   
mean                                 NaN      1.197519   
std                                  NaN      0.722372   
min                                  NaN      1.000000   
25%                                  NaN      1.000000   
50%                                  NaN      1.000000   
75%                                  NaN      1.000000   
max                                  NaN     21.000000   

                              product_id         price  \
count                              49980  4.998000e+04   
unique                             16864           NaN   
top     99a4788cb24856965c36a24e339b6058           NaN   
freq                                 366           NaN   
mean                                 NaN  2.607874e+06   
std                                  NaN  1.388282e+06   
min                                  NaN  2.000000e+05   
25%                                  NaN  1.411000e+06   
50%                                  NaN  2.610000e+06   
75%                                  NaN  3.810000e+06   
max                                  NaN  5.000000e+06   

                               seller_id  freight_value  \
count                              49980   49980.000000   
unique                              1777            NaN   
top     4a3ca9315b744ce9f8e9374361493884            NaN   
freq                                1236            NaN   
mean                                 NaN  104518.707483   
std                                  NaN   55180.931697   
min                                  NaN    9000.000000   
25%                                  NaN   57000.000000   
50%                                  NaN  104000.000000   
75%                                  NaN  152000.000000   
max                                  NaN  200000.000000   

                             customer_id order_status purchase_date  \
count                              49980        49980         49980   
unique                             42678            7           356   
top     fc3d1daec319d62d49bfb5e1f83123e9    delivered    24/11/2017   
freq                                  21        48645          1015   
mean                                 NaN          NaN           NaN   
std                                  NaN          NaN           NaN   
min                                  NaN          NaN           NaN   
25%                                  NaN          NaN           NaN   
50%                                  NaN          NaN           NaN   
75%                                  NaN          NaN           NaN   
max                                  NaN          NaN           NaN   

           payment_type product_category_name  product_weight_gram  
count             49980                 49980         49980.000000  
unique                4                     8                  NaN  
top     virtual account                  toys                  NaN  
freq              12579                  6373                  NaN  
mean                NaN                   NaN          2201.830892  
std                 NaN                   NaN          3929.896875  
min                 NaN                   NaN            50.000000  
25%                 NaN                   NaN           300.000000  
50%                 NaN                   NaN           800.000000  
75%                 NaN                   NaN          1850.000000  
max                 NaN                   NaN         40425.000000

Check for how much sales base on category

In [30]:
order_df.groupby(by="product_category_name").agg({
  "order_id": "nunique",
  "quantity" : "sum",
  "price": "sum"
})

order_id  quantity        price
product_category_name                                 
automotive                 6152      7501  16487385000
beauty                     6069      7448  16266276000
electronics                6124      7536  16248217000
fashion                    6101      7484  16427270000
gadget                     6071      7390  16055437000
sports                     6013      7406  16214246000
toys                       6204      7703  16599763000
utilities                  5956      7384  16042937000

As we can see automotive category bringing the most revenue with 16.487.385.000 but for most item sold by quantity or order we have toys category at 7703 item sold and 6204 item order.

Checking how much is freight value based on product category

In [34]:
order_df.groupby(by="product_category_name").agg({
  "order_id": "nunique",
  "freight_value" : ["sum", "mean"]
})

order_id freight_value               
                       nunique           sum           mean
product_category_name                                      
automotive                6152     658598000  104805.537874
beauty                    6069     644499000  103583.895853
electronics               6124     658224000  104513.178787
fashion                   6101     653698000  104224.808673
gadget                    6071     649926000  104540.131896
sports                    6013     649377000  104958.299661
toys                      6204     662482000  103951.357289
utilities                 5956     647041000  105604.863718

In [27]:
order_df.groupby(by="product_category_name").agg({
  "order_id": "nunique",
  "price": ["min", "max", "mean", "std"]
})

order_id   price                                     
                       nunique     min      max          mean           std
product_category_name                                                      
automotive                6152  200000  4999000  2.623709e+06  1.371808e+06
beauty                    6069  200000  5000000  2.614316e+06  1.393697e+06
electronics               6124  200000  4999000  2.579901e+06  1.387807e+06
fashion                   6101  201000  5000000  2.619144e+06  1.389720e+06
gadget                    6071  200000  5000000  2.582506e+06  1.397384e+06
sports                    6013  204000  4999000  2.620696e+06  1.387633e+06
toys                      6204  201000  4999000  2.604702e+06  1.394932e+06
utilities                 5956  201000  4999000  2.618400e+06  1.383042e+06